sdk/core/azure-core/inc/azure/core/internal/client_options.hpp

```diff
@@ -0,0 +1,123 @@
+
+#include <memory>
+#include <vector>
+
+namespace Azure { namespace Core { namespace Internal {
+
+  /**
+   * @brief  Base type for all client option types, exposes various common client options like Retry
+   * and Transport.
+   *
+   */
+  class ClientOptions {
+  private:
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perOperationPolicies;
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> m_perRetryPolicies;
+
+  public:
+    /**
+     * @brief Move each policy from \p options into the new instance.
+     *
+     */
+    explicit ClientOptions(ClientOptions&& options)
+        : m_perOperationPolicies(std::move(options.m_perOperationPolicies)),
+          m_perRetryPolicies(std::move(options.m_perRetryPolicies))
+    {
+    }
+
+    /**
+     * @brief Copy each policy to the new instance.
+     *
+     */
+    explicit ClientOptions(ClientOptions const& options)
+    {
+      m_perOperationPolicies.reserve(options.m_perOperationPolicies.size());
+      for (auto& policy : options.m_perOperationPolicies)
+      {
+        m_perOperationPolicies.emplace_back(policy->Clone());
+      }
+      m_perRetryPolicies.reserve(options.m_perRetryPolicies.size());
+      for (auto& policy : options.m_perRetryPolicies)
+      {
+        m_perRetryPolicies.emplace_back(policy->Clone());
+      }
+    }
+
+    ClientOptions() = default;
+
+    /**
+     * @brief Specify the number of retries and other retry-related options.
+     */
+    Azure::Core::Http::RetryOptions Retry;
+
+    /**
+     * @brief Customized HTTP client. We're going to use the default one if this is empty.
+     */
+    Azure::Core::Http::TransportOptions Transport;
+
+    /**
+     * @brief Telemetry options.
+     */
+    Azure::Core::Http::TelemetryOptions Telemetry;
+
+    /**
+     * @brief Adds a policy into the client.
+     *
+     * @remark The order of policy while sending the request is controlled by \p order.
+     * If you want the policy to execute once per client request use #HttpPolicyOrder::PerCall,
+     * otherwise use #HttpPolicyOrder::PerRetry to run the policy for every retry.
+     *
+     * @param policy The policy instance to be added to the pipeline.
+     * @param order The order to execute the policy.
+     */
+    void AddPolicy(
+        std::unique_ptr<Azure::Core::Http::HttpPolicy> policy,
+        Azure::Core::Http::HttpPolicyOrder order)
+    {
+      switch (order)
+      {
+        case Azure::Core::Http::HttpPolicyOrder::PerCall:
+          m_perOperationPolicies.push_back(std::move(policy));
+          break;
+        case Azure::Core::Http::HttpPolicyOrder::PerRetry:
+          m_perRetryPolicies.push_back(std::move(policy));
+          break;
+
+        default:
+          throw std::invalid_argument("Invalid order parameter");
+      }
+    }
+
+    /**
+     * @brief Get the Per Call Policies.
+     *
+     */
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> const& GetPerCallPolicies() const
```

Comment:
If you want encapsulation at all like you said in this comment https://github.com/Azure/azure-sdk-for-cpp/pull/1734/files#r582551593, why do you return const ref here? You expose the internal impl anyway

Reply:
This is internal,  (if you ask me, I would move it to be private and grant access to this function only to the httpPipeline class to keep it as a detail imp, but Rick prefer not to use friend access).

So, yes, this method is making a promise to the public about returning the list of policies as a vector<Policy> which we will never break.
It does not, however, make it part of the contract that the pipeline internally is a vector<Policy>.

This is not too bad because the `AddPolicy` takes a Policy, and you can get the list of policies we add with `GetPolicies`...

Reply:
> So, yes, this method is making a promise to the public about returning the list of policies as a vector which we will never break.

How? if as you said we change the internal impl from a vector to something else, we cannot return a const ref vector anymore for this function. This looks like a breaking change to me.

Reply:
>  (if you ask me, I would move it to be private and grant access to this function only to the httpPipeline class to keep it as a detail imp, but Rick prefer not to use friend access).

If you want to use private and friend class, what's the point of `AddPolicy` and `GetPolicies` anyway? I can just access the private members

Reply:
I admit I don't like this encapsulation idea in the first place. But if you insist, i'm fine with it. It just seems to me that your explanation of encapsulation doesn't hold water

Reply:
> How? if as you said we change the internal impl from a vector to something else, we cannot return a const ref vector anymore for this function. This looks like a breaking change to me.

The per call and per retry policies will be inside the clientOptions and will always be kept as a vector<Policy>.

The HttpPipeline will get them and use it to build a pipeline. The pipeline can be implemented as we want.  The clientOptions are inmutable.  The current implementation, for example, makes a clone of each policy from the clientOptions.  This ensure that the same clientOptions can be used to construct multiple clients, and each client will allocate its own pipeline.  Then we can dispose the clientOptions and the clients are still valid.
Then, for storage, a client can create a new client and re-use the same pipeline (shared_ptr).  The preferred way to create blobClients, for instance, is to spawn them from a containerClient instead of manually creating them from a clientOptions.

Reply:
okay, although I still don't think it's a good design. We cannot get anything if we keep arguing.